packages/0x.js/src/order_watcher/remaining_fillable_calculator.ts

```diff
@@ -0,0 +1,71 @@
+import {
+    SignedOrder,
+} from '../types';
+import { BigNumber } from 'bignumber.js';
+export class RemainingFillableCalculator {
+    private _signedOrder: SignedOrder;
+    private _isMakerTokenZRX: boolean;
+    private _transferrableMakerTokenAmount: BigNumber;
+    private _transferrableMakerFeeTokenAmount: BigNumber;
+    private _remainingMakerTokenAmount: BigNumber;
+    private _remainingMakerFeeAmount: BigNumber;
+    constructor(signedOrder: SignedOrder,
+                isMakerTokenZRX: boolean,
+                transferrableMakerTokenAmount: BigNumber,
+                transferrableMakerFeeTokenAmount: BigNumber,
+                remainingMakerTokenAmount: BigNumber) {
+        this._signedOrder = signedOrder;
+        this._isMakerTokenZRX = isMakerTokenZRX;
+        this._transferrableMakerTokenAmount = transferrableMakerTokenAmount;
+        this._transferrableMakerFeeTokenAmount = transferrableMakerFeeTokenAmount;
+        this._remainingMakerTokenAmount = remainingMakerTokenAmount;
+        this._remainingMakerFeeAmount = remainingMakerTokenAmount.times(signedOrder.makerFee)
+                                                                 .dividedToIntegerBy(signedOrder.makerTokenAmount);
+    }
+    public computeRemainingMakerFillable(): BigNumber {
+        if (this.hasSufficientFundsForFeeAndTransferAmount()) {
+            return this._remainingMakerTokenAmount;
+        }
+        if (this._signedOrder.makerFee.isZero()) {
+            return BigNumber.min(this._remainingMakerTokenAmount, this._transferrableMakerTokenAmount);
+        } else {
+            return this.calculatePartiallyFillableMakerTokenAmount();
+        }
+    }
+    public computeRemainingTakerFillable(): BigNumber {
+        return this.computeRemainingMakerFillable().times(this._signedOrder.takerTokenAmount)
+                                                   .dividedToIntegerBy(this._signedOrder.makerTokenAmount);
+    }
+    private hasSufficientFundsForFeeAndTransferAmount(): boolean {
+        if (this._isMakerTokenZRX) {
+            const totalZRXTransferAmountRequired = this._remainingMakerTokenAmount.plus(this._remainingMakerFeeAmount);
+            return this._transferrableMakerTokenAmount.gte(totalZRXTransferAmountRequired);
+        } else {
+            const hasSufficientFundsForTransferAmount = this._transferrableMakerTokenAmount.gte(
+                                                            this._remainingMakerTokenAmount);
+            const hasSufficientFundsForFeeAmount = this._transferrableMakerFeeTokenAmount.gte(
+                                                       this._remainingMakerFeeAmount);
+            return (hasSufficientFundsForTransferAmount && hasSufficientFundsForFeeAmount);
+        }
+    }
+
+    private calculatePartiallyFillableMakerTokenAmount(): BigNumber {
+        const orderToFeeRatio = this._signedOrder.makerTokenAmount.dividedToIntegerBy(this._signedOrder.makerFee);
```

Comment:
It's incorrect when fees are bigger than the order amount

Reply:
Moved to using non-integer division up until the end. Preventing error compounding and allowing Ratio < 1. Added a test for this as well.